src/ext/serializer.c

```diff
@@ -255,6 +378,101 @@ static void _add_assoc_zval_copy(zval *el, const char *name, zval *prop) {
+    do {                                                                        \
+        tmp = zend_hash_str_find(ht, key, sizeof(key) - 1);                     \
+        if (tmp) {                                                              \
+            if (Z_TYPE_P(tmp) != IS_STRING) {                                   \
+                /* zend_error(E_WARNING, "Value for %s is not string", key); */ \
+                smart_str_appends(str, "[unknown]");                            \
+            } else {                                                            \
+                smart_str_appends(str, Z_STRVAL_P(tmp));                        \
+            }                                                                   \
+        }                                                                       \
+    } while (0)
+
+/* This is modelled after _build_trace_string in PHP 7.0:
+ * @see https://github.com/php/php-src/blob/PHP-7.0/Zend/zend_exceptions.c#L581-L638
+ */
+static void _trace_string(smart_str *str, HashTable *ht, uint32_t num) /* {{{ */
+{
+    zval *file, *tmp;
+
+    smart_str_appendc(str, '#');
+    smart_str_append_long(str, num);
+    smart_str_appendc(str, ' ');
+
+    file = zend_hash_str_find(ht, "file", sizeof("file") - 1);
+    if (file) {
+        if (Z_TYPE_P(file) != IS_STRING) {
+            ddtrace_log_debug("serializer stack trace: Function name is not a string");
+            smart_str_appends(str, "[unknown function]");
+        } else {
+            zend_long line;
+            tmp = zend_hash_str_find(ht, "line", sizeof("line") - 1);
+            if (tmp) {
+                if (Z_TYPE_P(tmp) == IS_LONG) {
+                    line = Z_LVAL_P(tmp);
+                } else {
+                    ddtrace_log_debug("serializer stack trace: Line is not a long");
+                    line = 0;
+                }
+            } else {
+                line = 0;
+            }
+            smart_str_append(str, Z_STR_P(file));
+            smart_str_appendc(str, '(');
+            smart_str_append_long(str, line);
+            smart_str_appends(str, "): ");
+        }
+    } else {
+        smart_str_appends(str, "[internal function]: ");
+    }
+    _DD_TRACE_APPEND_KEY("class");
+    _DD_TRACE_APPEND_KEY("type");
+    _DD_TRACE_APPEND_KEY("function");
+    tmp = zend_hash_str_find(ht, "args", sizeof("args") - 1);
+
+    /* If there were arguments, show an ellipsis, otherwise nothing */
+    if (tmp && Z_TYPE_P(tmp) == IS_ARRAY && zend_hash_num_elements(Z_ARRVAL_P(tmp))) {
+        smart_str_appends(str, "(...)\n");
+    } else {
+        smart_str_appends(str, "()\n");
+    }
+}
+
+/* Modelled after getTraceAsString from PHP 5.4:
+ * @see https://lxr.room11.org/xref/php-src%405.4/Zend/zend_exceptions.c#609-635
+ */
+static void _serialize_stack_trace(zval *meta, zval *trace) {
+    zval *frame, output;
+    zend_ulong index;
```

Comment:
In one of my containers based on the `php:7.2-apache` image, I got this compiler warning:

```
/src/ddtrace/src/ext/serializer.c: In function '_serialize_stack_trace':
/src/ddtrace/src/ext/serializer.c:450:16: warning: variable 'index' set but not used [-Wunused-but-set-variable]
     zend_ulong index;
                ^~~~~
```

But it doesn't appear to be happening in our build containers. Not sure why it's showing that since `ZEND_HASH_FOREACH_NUM_KEY_VAL` uses `index`... weird. I think it's fine but just wanted to mention it just in case it shows up later in the CI builds. :)

Reply:
Actually, I compiled on the wrong branch, this is showing up in the centos containers

```bash
$ docker-compose run --rm 7.0-centos bash
```

And then:

```bash
$ yum -y install libcurl-devel && \
make all CFLAGS="-std=gnu11 -O2 -Wall -Wextra -Werror" ECHO_ARG="-e"
[ ... ]
/home/circleci/app/tmp/build_extension/src/ext/serializer.c: In function ‘_serialize_stack_trace’:
/home/circleci/app/tmp/build_extension/src/ext/serializer.c:450:16: error: variable ‘index’ set but not used [-Werror=unused-but-set-variable]
     zend_ulong index;
                ^~~~~
```

Reply:
It's not actually used because the variadic debug log is commented out. I can use a different FOREACH macro, or we can wait to do something if -Werror causes an issue. What do you think?

Reply:
I changed the macro I used -- should be good to go now.